### PR TITLE
mon/MonClient: random all ranks then pick first_n

### PR DESCRIPTION
**Pull request by LiumxNL**
Signed-off-by: Mingxin Liu <mingxin@xsky.com>

### CODE REVIEW

**Comment by LiumxNL** [comment]
@tchaikov please review, thanks!

File: src/mon/MonClient.cc [hunk 1]
```diff
@@ -625,17 +625,21 @@ MonConnection& MonClient::_add_conn(unsigned rank)
...
+  vector<unsigned> ranks(m);
+  for (unsigned i = 0; i < m; i++) {
     ranks[i] = i;
   }
   std::random_device rd;
   std::mt19937 rng(rd());
   std::shuffle(ranks.begin(), ranks.end(), rng);
+
+  unsigned n = cct->_conf->mon_client_hunt_parallel;
+  if (n) {
```
Comment (tchaikov) [inline src/mon/MonClient.cc]: could you keep the original block of `if (n == 0 || n > monmap.size())`, IMHO, it's more explicit and easier to digest

File: src/mon/MonClient.cc [hunk 2]
```diff
@@ -625,17 +625,21 @@ MonConnection& MonClient::_add_conn(unsigned rank)
 
 void MonClient::_add_conns()
 {
-  unsigned n = cct->_conf->mon_client_hunt_parallel;
-  if (n == 0 || n > monmap.size()) {
-    n = monmap.size();
-  }
-  vector<unsigned> ranks(n);
-  for (unsigned i = 0; i < n; i++) {
+  unsigned m = monmap.size();
```
Comment (tchaikov) [inline src/mon/MonClient.cc]: nit, could have a better name than `m`, like `num_mons`, also could make this `const`.

**Comment by liewegas** [comment]
I'm hitting this bug from rados suite: http://pulpito.ceph.com/sage-2017-02-17_15:21:53-rados:multimon-wip-sage-testing---basic-smithi

Including this in my test branch!

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-02-17_22:18:26-rados:multimon-wip-sage-testing---basic-smithi/

it works!

**Comment by LiumxNL** [comment]
@tchaikov updated.

**Pull request merged**